[AArch64][SME2] Improve register allocation of multi-vector SME intrinsics

kmclaughlin-arm opened the pull request:
The FORM_TRANSPOSED_REG_TUPLE pseudos have been created to
improve register allocation for intrinsics which use strided and
contiguous multi-vector registers, avoiding unnecessary copies.

If the operands of the pseudo are copies where the source register is in
the StridedOrContiguous class, the pseudo is used by getRegAllocationHints
to suggest a contigious multi-vector register which matches the subregister
sequence used by the operands.
If the operands do not match this pattern, the pseudos are expanded
to a REG_SEQUENCE.

Patch contains changes by Matthew Devereau.

kmclaughlin-arm committed:
[AArch64][SME2] Add tests for strided load & contiguous dot instructions

kmclaughlin-arm committed:
Enable subreg liveness in SME2 dot tests.

kmclaughlin-arm committed:
[AArch64][SME2] Add FORM_STRIDED_TUPLE pseudo nodes

This patch adds a pseudo node to help towards improving register
allocation of multi-vector SME intrinsics.

The FORM_STRIDED_TUPLE node is emitted if each of the operands of a
contiguous multi-vector dot intrinsic are the result of a strided
multi-vector load. The operands of the psuedo will be one subregister
at the same index from each of these strided loads.

Follow up patches will use this pseudo when adding register allocation
hints to remove unecessary register copies in this scenario. Subregister
liveness is also required to achieve this and has been enabled in the
tests changed by this patch.

Patch contains changes by Matthew Devereau.

kmclaughlin-arm committed:
[AArch64][SME2] Add register allocation hints for ZPRMulReg

This patch implements getRegAllocationHints to improve register
allocation for the ZPR2Mul2Reg & ZPR4Mul4Reg classes.
If a FORM_STRIDED_TUPLE is found, getRegAllocationHints will try
to find a contiguous ZPRMulReg beginning with the same subregister
as the first operand of the pseudo.

For example, if the first strided load has been assigned $z16_z20_z24_z28
and the operands of the pseudo are each accessing subregister zsub2, the
correct register to use would be $z24_z25_z26_z27.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -1142,6 +1146,30 @@ bool AArch64ExpandPseudo::expandMultiVecPseudo(
+    Register FormTupleOpReg = MI.getOperand(i + 1).getReg();
+    Register ReturnTupleSubReg =
+        TRI->getSubReg(ReturnTuple, AArch64::zsub0 + i);
+    if (FormTupleOpReg != ReturnTupleSubReg)
```
I think a comment about why this check is necessary would be good.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5709,6 +5711,46 @@ SDValue AArch64TargetLowering::getRuntimePStateSM(SelectionDAG &DAG,
 
+static unsigned getIntrinsicID(const SDNode *N);
+
+SDValue TryLowerMultiVecSMEDotIntrinsic(SDValue Op, SelectionDAG &DAG,
```
Try -> try

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5709,6 +5711,46 @@ SDValue AArch64TargetLowering::getRuntimePStateSM(SelectionDAG &DAG,
                      Mask);
 }
 
+static unsigned getIntrinsicID(const SDNode *N);
```
Doesn't this already exist? It's used in other places in this file.

MacDue replied:
It's not defined until a few K lines below this

kmclaughlin-arm committed:
- Removed the FORM_STRIDED_TUPLE_X# nodes, leaving only the pseudo nodes.
- Changed the tablegen patterns used by the dot intrinsics to always output
  the FORM_STRIDED_TUPLE_X#_PSEUDO nodes.
- Check that the operands to the pseudo are copies from a StridedOrContiguous
  register class in AdjustInstrPostInstrSelection, falling back on creating
  a REG_SEQUENCE node if not.

MacDue commented on llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td:
```diff
@@ -28,6 +28,17 @@ def AArch64_restore_zt : SDNode<"AArch64ISD::RESTORE_ZT", SDTypeProfile<0, 2,
                              [SDTCisInt<0>, SDTCisPtrTy<1>]>,
                              [SDNPHasChain, SDNPSideEffect, SDNPMayStore]>;
+
+def SDT_FORM_STRIDED_TUPLE_X2 : SDTypeProfile<4, 4,
```
Might be wrong, but shouldn't this be two results and two operands?
```suggestion
def SDT_FORM_STRIDED_TUPLE_X2 : SDTypeProfile<2, 2,
```

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7438,3 +7448,20 @@ bool AArch64DAGToDAGISel::SelectSMETileSlice(SDValue N, unsigned MaxSize,
+void AArch64DAGToDAGISel::SelectFormTuplePseudo(SDNode *Node, unsigned Size) {
+  assert((Size == 2 || Size == 4) && "Invalid Tuple size");
+  EVT VT = Node->getValueType(0);
+  SmallVector<SDValue> Ops;
+  for (unsigned I = 0; I < Size; I++)
+    Ops.push_back(Node->getOperand(I));
```
nit:
```suggestion
  SmallVector<SDValue> Ops(Node->op_values());
```

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -7438,3 +7448,20 @@ bool AArch64DAGToDAGISel::SelectSMETileSlice(SDValue N, unsigned MaxSize,
   return true;
 }
+
+void AArch64DAGToDAGISel::SelectFormTuplePseudo(SDNode *Node, unsigned Size) {
+  assert((Size == 2 || Size == 4) && "Invalid Tuple size");
```
Isn't `Size` just `Node->getNumOperands()` (or `getNumValues()`)?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -5898,6 +5940,22 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_VOID(SDValue Op,
         Op->getOperand(0), // Chain
         DAG.getTargetConstant((int32_t)(AArch64SVCR::SVCRZA), DL, MVT::i32),
         DAG.getConstant(AArch64SME::Always, DL, MVT::i64));
+  case Intrinsic::aarch64_sme_uvdot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_suvdot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_usvdot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_svdot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_usdot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_udot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_sudot_lane_za32_vg1x4:
+  case Intrinsic::aarch64_sme_sdot_lane_za32_vg1x4:
+    return TryLowerMultiVecSMEDotIntrinsic(Op, DAG, 4);
+  case Intrinsic::aarch64_sme_uvdot_lane_za32_vg1x2:
+  case Intrinsic::aarch64_sme_sdot_lane_za32_vg1x2:
+  case Intrinsic::aarch64_sme_svdot_lane_za32_vg1x2:
+  case Intrinsic::aarch64_sme_usdot_lane_za32_vg1x2:
+  case Intrinsic::aarch64_sme_sudot_lane_za32_vg1x2:
+  case Intrinsic::aarch64_sme_udot_lane_za32_vg1x2:
+    return TryLowerMultiVecSMEDotIntrinsic(Op, DAG, 2);
```
Perhaps a structurally simpler way to implement this (avoiding the need to do custom isel) is to change the patterns to always use the `FORM_STRIDED_TUPLE_X.._PSEUDO` instruction instead of `REG_SEQUENCE`.

For the multi-vector load case that you're trying to improve, the inputs to the tuple are always COPY nodes of the form:

```
%9:zpr = COPY %7.zsub0:zpr2stridedorcontiguous
```

There are cases where the RegisterCoalescer can make better decisions when using regular COPY nodes rather than the FORM_STRIDED_TUPLE pseudos. We could choose to handle the FORM_STRIDED_TUPLE pseudo with the `hasPostISelHook = 1` where directly post-isel they are transformed into a REG_SEQUENCE node when any of the input values are not COPY nodes where the source register is in  a 'stridedorcontiguous' register class. The REG_SEQUENCE node itself is then lowered later by the TwoAddressInstructionPass into individual COPY nodes.

kmclaughlin-arm replied:
Thank you for the suggestion, @sdesmalen-arm. I was able to remove the `FORM_STRIDED_TUPLE` nodes and instead add `hasPostISelHook = 1` to the pseudos, creating a `REG_SEQUENCE` if the input values are not copies from a StridedOrContiguous source register.
This has been added in a new commit, with the RegAllocHints commit added on top.

sdesmalen-arm reviewed (commented):
This PR doesn't make much sense without seeing the accompanying PR that improves the register allocation using these nodes. Could you either fold those changes into this PR, or stack another PR on top of this one?

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp:
```diff
@@ -504,6 +504,8 @@ class AArch64DAGToDAGISel : public SelectionDAGISel {
 
   bool SelectAllActivePredicate(SDValue N);
   bool SelectAnyPredicate(SDValue N);
+
+  void SelectFormTuplePseudo(SDNode *N);
```
nit: This is unused now

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -2808,6 +2808,8 @@ const char *AArch64TargetLowering::getTargetNodeName(unsigned Opcode) const {
     MAKE_CASE(AArch64ISD::FMUL_PRED)
     MAKE_CASE(AArch64ISD::FSUB_PRED)
     MAKE_CASE(AArch64ISD::RDSVL)
+    MAKE_CASE(AArch64ISD::FORM_STRIDED_TUPLE_X2)
+    MAKE_CASE(AArch64ISD::FORM_STRIDED_TUPLE_X4)
```
Unused?

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.h:
```diff
@@ -478,6 +478,9 @@ enum NodeType : unsigned {
   SME_ZA_LDR,
   SME_ZA_STR,
 
+  FORM_STRIDED_TUPLE_X2,
+  FORM_STRIDED_TUPLE_X4,
+
```
Unused?

MacDue reviewed (commented):
This remaining FORM_STRIDED_TUPLE_X2/4 code looks like it's probably unused:

MDevereau commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,81 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+    if (!VRM->hasPhys(FirstLoadVirtReg))
+      continue;
+
+    unsigned SubRegIdx = 0;
+    MCRegister FirstLoadPhysReg = VRM->getPhys(FirstLoadVirtReg);
+
+    // The subreg number is used to access the correct unit of the
+    // strided register found in the map above.
+    switch (MI.getOperand(1).getSubReg()) {
+    case AArch64::zsub0:
+      break;
+    case AArch64::zsub1:
+      SubRegIdx = 1;
+      break;
+    case AArch64::zsub2:
+      SubRegIdx = 2;
+      break;
+    case AArch64::zsub3:
+      SubRegIdx = 3;
+      break;
+    default:
+      continue;
+    }
```
Is it possible to shorten this to
```c++
unsigned SubRegIdx = AArch64::zsub0 - MI.getOperand(1).getSubReg();
if (SubRegIdx > 3)
  continue;
```
?

kmclaughlin-arm replied:
In the latest commit I changed SubRegIdx to a signed int. The reason was in case `getSubReg()` happens to return a register number which is smaller than `AArch64::zsub0`.

kmclaughlin-arm committed:
- Remove unused references to FORM_STRIDED_TUPLE
- Removed switch from getRegAllocationHints

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8666,6 +8671,77 @@ void AArch64TargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
+          Size == 2 ? &AArch64::ZPR2StridedOrContiguousRegClass
+                    : &AArch64::ZPR4StridedOrContiguousRegClass;
+
+      if (MRI.getRegClass(Ld->getReg()) == RegClass)
+        UseFormStrided = true;
```
It would help to have this loop in a separate function that returns `false` if there is any reason that the operands don't meet the criteria. Then you can also include the check for the subreg indices in the same loop (for which you currently require a `std::equal`).

kmclaughlin-arm replied:
I've moved this to a new function called `shouldUseFormStridedPseudo()` & removed the `std::equal` test.

kmclaughlin-arm committed:
- Removed Multi_Index_FormStrided_Pat classes and reused existing classes
- Removed isReg() check for operand 1 of a copy
- Moved loop over FORM_STRIDED_TUPLE operands to new function.

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8666,6 +8671,77 @@ void AArch64TargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
+    SmallVector<unsigned, 4> OpSubRegs;
+    for (unsigned I = 1; I < MI.getNumOperands(); ++I) {
+      MachineOperand &MO = MI.getOperand(I);
+      if (!MO.isReg())
+        continue;
+
```
That is right, I've changed this to an assert.

kmclaughlin-arm commented on llvm/lib/Target/AArch64/SMEInstrFormats.td:
```diff
@@ -174,6 +188,12 @@ class SME2_ZA_TwoOp_VG2_Multi_Index_Pat<string name, SDPatternOperator intrinsic
+                                        Operand imm_ty, ComplexPattern tileslice>
+    : Pat<(intrinsic (i32 (tileslice MatrixIndexGPR32Op8_11:$base, index_ty:$offset)), vt:$Zn1, vt:$Zn2, vt:$Zm, (i32 imm_ty:$i)),
+          (!cast<Instruction>(name # _PSEUDO) $base, $offset,
+                                              (FORM_STRIDED_TUPLE_X2_PSEUDO vt:$Zn1,vt:$Zn2), zpr_ty:$Zm, imm_ty:$i)>;
```
There are a number of other intrinsics which use these patterns other than sdot/udot, which is why I initially added a new pattern. However, since we will fall back on REG_SEQUENCE anyway if the expected copy instructions are not found I think we can just update the existing one.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+    if (SubRegIdx < 0 || SubRegIdx > 3)
+      continue;
+
+    SmallVector<Register, 4> RegUnits;
+    for (MCRegUnit Unit : TRI->regunits(FirstLoadPhysReg))
+      RegUnits.push_back(Unit);
+
+    // Find the contiguous ZPRMul register which starts with the
+    // same register unit as the strided register and add to Hints.
+    Register StartReg = RegUnits[SubRegIdx];
+    for (unsigned I = 0; I < Order.size(); ++I) {
+      Register Reg = *TRI->regunits(Order[I]).begin();
+      if (Reg == StartReg)
+        Hints.push_back(Order[I]);
+    }
```
There is no need to iterate through all MCRegUnits for this register. It also feels rather fiddly to index into `RegUnits[SubRegIdx], because it makes assumptions on the order of register units in `RegUnits`.

You can do this instead using `getSubReg`, e.g.
```
MCRegister TupleStartReg = getSubReg(VRM->getPhys(FirstLoadVirtReg), MI.getOperand(1).getSubReg());
for (unsigned I = 0; I < Order.size(); ++I) 
  if (MCRegister R = getSubReg(Order[I], AArch64::zsub0))
    if (R == TupleStartReg)
      ....
}
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+      continue;
+
+    // Look up the physical register mapped to the first load of the pseudo.
+    Register FirstLoadVirtReg = MI.getOperand(1).getReg();
```
Please don't name variables with the assumption that these result from certain operations, like Loads.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+
+    // The subreg number is used to access the correct unit of the
+    // strided register found in the map above.
+    SubRegIdx = MI.getOperand(1).getSubReg() - AArch64::zsub0;
+    if (SubRegIdx < 0 || SubRegIdx > 3)
+      continue;
```
Please use a switch statement, such that explicitly only zsub0..zsub3 are supported.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+      TargetRegisterInfo::getRegAllocationHints(VirtReg, Order, Hints, MF, VRM);
+
+  unsigned RegID = MRI.getRegClass(VirtReg)->getID();
+  if (RegID != AArch64::ZPR2Mul2RegClassID &&
+      RegID != AArch64::ZPR4Mul4RegClassID)
+    return DefaultHints;
+
```
I think this condition can be removed (it would e.g. be equally valid for other register class with contiguous tuples, or perhaps just any register class?).

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+    SmallVectorImpl<MCPhysReg> &Hints, const MachineFunction &MF,
+    const VirtRegMap *VRM, const LiveRegMatrix *Matrix) const {
+  const AArch64Subtarget &STI = MF.getSubtarget<AArch64Subtarget>();
+  const TargetRegisterInfo *TRI = STI.getRegisterInfo();
```
Can be removed because `AArch64RegisterInfo` `this` == `TRI`.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,69 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+  const AArch64Subtarget &STI = MF.getSubtarget<AArch64Subtarget>();
+  const TargetRegisterInfo *TRI = STI.getRegisterInfo();
+  const MachineRegisterInfo &MRI = MF.getRegInfo();
+  bool DefaultHints =
+      TargetRegisterInfo::getRegAllocationHints(VirtReg, Order, Hints, MF, VRM);
```
We should prioritise the tuples as added below over any generic hints. That means this function should be called last.

kmclaughlin-arm committed:
- shouldUseFormStridedPseudo sets RegClass using a switch statement
- Removed INTRINSIC_W_CHAIN from getIntrinsicID
- Renamed FORM_STRIDED_TUPLE_X#_PSEUDO -> FORM_TRANSPOSED_REG_TUPLE_X#_PSEUDO
- Add switch statement back into getRegAllocationHints
- Use getSubReg in getRegAllocationHints and remove index into RegUnits

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8641,6 +8646,55 @@ static bool checkZExtBool(SDValue Arg, const SelectionDAG &DAG) {
+  MachineRegisterInfo &MRI = MI.getMF()->getRegInfo();
+  bool UseFormStrided = false;
+  unsigned NumOperands =
+      MI.getOpcode() == AArch64::FORM_STRIDED_TUPLE_X2_PSEUDO ? 2 : 4;
```
Renamed to `FORM_TRANSPOSED_REG_TUPLE_X2_PSEUDO` & `AArch64::FORM_TRANSPOSED_REG_TUPLE_X4_PSEUDO`

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8641,6 +8641,57 @@ static bool checkZExtBool(SDValue Arg, const SelectionDAG &DAG) {
+      return false;
+
+    const TargetRegisterClass *RegClass = nullptr;
+    switch (MI.getNumOperands() - 1) {
```
Could you hoist this out of the loop, and also just switch on the opcode, rather than the number of operands? (then you can add a `llvm_unreachable` for the case where the opcode is unexpected).

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8641,6 +8641,57 @@ static bool checkZExtBool(SDValue Arg, const SelectionDAG &DAG) {
+      return false;
+
+    const MachineOperand &CpySrc = Def->getParent()->getOperand(1);
+    MachineOperand *CopySrcOp = MRI.getOneDef(CpySrc.getReg());
```
nit: move closer to def.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8641,6 +8641,57 @@ static bool checkZExtBool(SDValue Arg, const SelectionDAG &DAG) {
+    if (!Def || !Def->getParent()->isCopy())
+      return false;
+
+    const MachineOperand &CpySrc = Def->getParent()->getOperand(1);
```
nit: 
```suggestion
    const MachineOperand &CopySrc = Def->getParent()->getOperand(1);
```
(to be consistent with the 'o' in Copy you've added to the other variable name)

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8666,6 +8717,29 @@ void AArch64TargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
     }
   }
 
+  if (MI.getOpcode() == AArch64::FORM_TRANSPOSED_REG_TUPLE_X2_PSEUDO ||
+      MI.getOpcode() == AArch64::FORM_TRANSPOSED_REG_TUPLE_X4_PSEUDO) {
+    // If input values to the FORM_TRANSPOSED_REG_TUPLE pseudo aren't copies
+    // from a StridedOrContiguous class, fall back on REG_SEQUENCE node.
+    if (!shouldUseFormStridedPseudo(MI)) {
```
nit: Maybe fold those conditions into one? or otherwise bail out early if `shouldUseFormStridedPseudo(MI) == true`, to avoid a level of indentation?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,56 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+        MI.getOpcode() != AArch64::FORM_TRANSPOSED_REG_TUPLE_X4_PSEUDO)
+      continue;
+
+    switch (MI.getOperand(1).getSubReg()) {
```
nit: please make a variable for `MI.getOperand(1).getsubReg()`, because it's used more than once.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -1107,6 +1107,56 @@ unsigned AArch64RegisterInfo::getRegPressureLimit(const TargetRegisterClass *RC,
+  for (MachineInstr &MI : MRI.def_instructions(VirtReg)) {
+    if (MI.getOpcode() != AArch64::FORM_TRANSPOSED_REG_TUPLE_X2_PSEUDO &&
+        MI.getOpcode() != AArch64::FORM_TRANSPOSED_REG_TUPLE_X4_PSEUDO)
+      continue;
```
nit: I wonder if we just want to bail out early at this point?

sdesmalen-arm commented on llvm/lib/Target/AArch64/SMEInstrFormats.td:
```diff
@@ -34,6 +34,20 @@ def tileslicerange0s4 : ComplexPattern<i32, 2, "SelectSMETileSlice<0,  4>", []>;
 
 def am_sme_indexed_b4 :ComplexPattern<iPTR, 2, "SelectAddrModeIndexedSVE<0,15>", [], [SDNPWantRoot]>;
 
+def FORM_TRANSPOSED_REG_TUPLE_X2_PSEUDO :
```
I think this needs a description of why we add these pseudos, and a comment that we expand them to REG_SEQUENCE with the post-isel hook if they don't meet certain criteria.

kmclaughlin-arm committed:
- Renamed CpySrc -> CopySrc
- Moved switch in shouldUseFormStridedPseudo outside of loop
- Added a description of the pseudo nodes to SMEInstrFormats.td

kmclaughlin-arm committed:
- Return early from getRegAllocationHints if the opcode is not a FORM_TRANSPOSED
  pseudo or the subreg indices are not in the range zsub0-zsub3.

kmclaughlin-arm committed:
- Return default hints getRegAllocationHints if opcodes don't match

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -1142,6 +1146,32 @@ bool AArch64ExpandPseudo::expandMultiVecPseudo(
+
+  const TargetRegisterInfo *TRI =
+      MBB.getParent()->getSubtarget().getRegisterInfo();
+  for (unsigned i = 0; i < Size; i++) {
```
nit: 
```suggestion
  for (unsigned I = 0; I < Size; ++I) {
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8641,6 +8641,58 @@ static bool checkZExtBool(SDValue Arg, const SelectionDAG &DAG) {
+      SubReg = OpSubReg;
+
+    MachineOperand *CopySrcOp = MRI.getOneDef(CopySrc.getReg());
+    if (!CopySrcOp || !CopySrcOp->isReg() || OpSubReg != SubReg)
+      return false;
+
+    if (MRI.getRegClass(CopySrcOp->getReg()) != RegClass)
+      return false;
```
nit:
```suggestion
    if (!CopySrcOp || !CopySrcOp->isReg() || OpSubReg != SubReg ||
        MRI.getRegClass(CopySrcOp->getReg()) != RegClass)
      return false;
```

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8666,6 +8718,30 @@ void AArch64TargetLowering::AdjustInstrPostInstrSelection(MachineInstr &MI,
+
+    for (unsigned I = 1; I < MI.getNumOperands(); ++I) {
+      MIB.add(MI.getOperand(I));
+      MIB.addImm(SubRegs[I - 1]);
```
nit:
```suggestion
      MIB.addImm(AArch64::zsub0 + (I-1));
```
Then you can remove `SubRegs[]`.

sdesmalen-arm reviewed (approved):
Some minor nits, but otherwise LGTM.

Could you also please update the PR's title and commit message, to reflect the new name before committing?

kmclaughlin-arm committed:
- Removed SubRegs list from AdjustInstrPostInstrSelection
